clang/lib/Driver/ToolChains/SYCL.cpp

```diff
@@ -203,8 +233,17 @@ const char *SYCL::Linker::constructLLVMLinkCommand(
     for (const auto &II : InputFiles) {
       std::string FileName = getToolChain().getInputFilename(II);
       if (II.getType() == types::TY_Tempfilelist) {
-        // Pass the unbundled list with '@' to be processed.
-        Libs.push_back(C.getArgs().MakeArgString("@" + FileName));
+        if (IsRDC) {
+          // Pass the unbundled list with '@' to be processed.
+          Libs.push_back(C.getArgs().MakeArgString("@" + FileName));
+        } else {
+          assert(InputFiles.size() == 2 &&
+                 "Unexpected inputs for no-RDC with temp file list");
+          // If we're in no-RDC mode and the input is a temp file list,
+          // we want to link multiple object files each against device libs,
+          // so we should consider this input as an object and not pass '@'.
```

Comment:
If we are considering the input as a temp file list, it should be treated as a list.  Maybe the input type should be updated to reflect the actual type of file?

Reply:
@mdtoguchi 
I might need your help coming up with a solution for this one, so here's what happens. 

This is the case of fat static archive.

So we have:
1) spirv-to-ir-wrapper, on fat archive, outputs temp file list of all device code objects in it
2) link all sycl libs together into a single bc, no device code involved here
3) link each entry in 1) with 2) 

So action 3 has two inputs, the temp file list containing LLVM BC from 1) and the single LLVM BC from 2)

To get 3) to happen per-entry in temp file list, I use ForEachWrappingAction

When ForEachWrappingAction is processed, it creates the job for whatever you want to foreach (here, from the linker action in 3) first. This is where SYCL::Linker::constructLLVMLinkCommand is called from and where that code is hit. 

After that job is created (as well as any inputs jobs), ForEachWrappingAction steals all of the jobs out of the main list that were just created and decomposes their inputs into the constructLLVMForeachCommand call and adds the newly created llvm-foreach job to the list.

So when constructLLVMLinkCommand is running, it has no idea it's being used in a foreach, at this point it only sees the input from 1) which is a temp file list, but since it will be used in a llvm-foreach where the temp file list input arg is replaced by the llvm-foreach tool to be each entry in the temp file list, we don't want to use '@' since it's really a single entry.

I can't think of a clean way to have constructLLVMLinkCommand get the type to be LLVM BC (matching the real input to llvm-link at runtime) 

In the general case, I don't think we necessarily even know what the file type inside the temp file list is. I can't even think of a clean way to only handle it for linker jobs inside llvm-foreach. At action creation time, we do know it's used in foreach, so maybe we could do something there and add information to LinkJobAction saying which inputs to it are really a foreach?

Reply:
Unfortunately, I can't think of a clean way to approach this either. Given there is no expectation of any change to the toolchain when using ForEachWrappingAction, we don't have a lot of control over the inputs.  As long as we are guaranteed that any inputs to LLVMLink with NoRDC with the temp file list type is a true file we may have to stick with what you have.

Reply:
Yeah, at the moment that is the case. If we ever need to use '@' in the link command with no-rdc we will need to come up with something, but we never need to do that today.